src/core/agent/services/ipexCommunicationService.ts

```diff
@@ -171,59 +170,153 @@ class IpexCommunicationService extends AgentService {
+  async offerAcdcFromApply(id: string, acdc?: any) {
+    const applyNoteRecord = await this.notificationStorage.findById(id);
+
+    if (!applyNoteRecord) {
+      throw new Error(
+        `${IpexCommunicationService.NOTIFICATION_NOT_FOUND} ${id}`
+      );
+    }
+
+    if (Object.keys(applyNoteRecord.linkedGroupRequests).length) {
+      for (const said of Object.keys(applyNoteRecord.linkedGroupRequests)) {
+        if (!applyNoteRecord.linkedGroupRequests[said]) {
+          await this.offerAcdcFromMultisigExn(said as string);
+        }
+      }
+      return;
+    }
 
+    const msgSaid = applyNoteRecord.a.d as string;
+    const applyExn = await this.props.signifyClient.exchanges().get(msgSaid);
+    const holder = await this.identifierStorage.getIdentifierMetadata(
+      applyExn.exn.a.i
+    );
     const holderSignifyName = (
-      await this.identifierStorage.getIdentifierMetadata(msg.exn.a.i)
+      await this.identifierStorage.getIdentifierMetadata(applyExn.exn.a.i)
     ).signifyName;
 
-    const [offer, sigs, end] = await this.props.signifyClient.ipex().offer({
-      senderName: holderSignifyName,
-      recipient: msg.exn.i,
-      acdc: new Serder(acdc),
-      applySaid: msg.exn.d,
+    let op: Operation;
+    if (holder.multisigManageAid) {
+      const { op: opMultisigOffer, exnSaid } =
+        await this.multisigService.offerPresentMultisigACDC(
+          holderSignifyName,
+          msgSaid,
+          acdc,
+          applyExn.exn.i
+        );
+      op = opMultisigOffer;
+      applyNoteRecord.linkedGroupRequests = {
+        [exnSaid]: true,
+      };
+      await this.notificationStorage.update(applyNoteRecord);
+    } else {
+      const [offer, sigs, end] = await this.props.signifyClient.ipex().offer({
+        senderName: holderSignifyName,
+        recipient: applyExn.exn.i,
+        acdc: new Serder(acdc),
+        applySaid: applyExn.exn.d,
+      });
+      op = await this.props.signifyClient
+        .ipex()
+        .submitOffer(holderSignifyName, offer, sigs, end, [applyExn.exn.i]);
+    }
+
+    const pendingOperation = await this.operationPendingStorage.save({
+      id: op.name,
+      recordType: OperationPendingRecordType.ExchangeOfferPresentCredential,
     });
-    await this.props.signifyClient
-      .ipex()
-      .submitOffer(holderSignifyName, offer, sigs, end, [msg.exn.i]);
-    await Agent.agent.keriaNotifications.deleteNotificationRecordById(
-      notification.id,
-      notification.a.r as NotificationRoute
-    );
+
+    Agent.agent.keriaNotifications.addPendingOperationToQueue(pendingOperation);
+    if (!holder.multisigManageAid) {
+      await Agent.agent.keriaNotifications.deleteNotificationRecordById(
+        id,
+        applyNoteRecord.a.r as NotificationRoute
+      );
+    }
   }
 
   @OnlineOnly
-  async grantAcdcFromAgree(msgSaid: string) {
-    const msgAgree = await this.props.signifyClient.exchanges().get(msgSaid);
+  async grantAcdcFromAgree(id: string) {
```

Comment:
same here, can we just pass the entire notification instead of fetching it again?